src/exclusion/object_filter.hpp

```diff
@@ -0,0 +1,73 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#pragma once
+
+#include <map>
+#include <set>
+#include <stack>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include <clock.hpp>
+#include <config.hpp>
+#include <log.hpp>
+#include <manifest.hpp>
+#include <object_store.hpp>
+
+namespace ddwaf::exclusion {
+
+class path_trie {
+public:
+    path_trie() = default;
+
+    [[nodiscard]] path_trie find(std::string_view key) const;
+
+    template <typename T> [[nodiscard]] path_trie find(const std::vector<T> &path) const;
+
+    template <typename T> void insert(const std::vector<T> &path);
+
+    [[nodiscard]] bool is_terminal() const { return root && root->terminal; }
+    [[nodiscard]] bool is_valid() const { return root && !root->values.empty(); }
+
+protected:
+    struct trie_node {
+        // Heterogenous lookups required...
+        std::unordered_map<std::string_view, std::shared_ptr<trie_node>> values;
+        // TODO this shouldn't be necessary
+        bool terminal{false};
+    };
+
+    explicit path_trie(std::shared_ptr<trie_node> node_) : root(std::move(node_)) {}
+
+    std::string_view get_stored_string(std::string_view str);
+
+    std::shared_ptr<trie_node> root{nullptr};
```

Comment:
It needs to be shared so that we can create a `path_trie` from a `trie_node` contained in the parent `path_trie`.

Reply:
My other comment was lost. I got that eventually, that you have these find methods returning subtries. What I had written was that the way the trie was written did not help comprehension because the implementation bleeds into an aux function of object_filter::match. It would be better if the trie only exposed an `add` and a `bool contains()` function. You could then use unique_ptr's. Also you could implement node compression.

Reply:
The reason I implemented the `path_trie` in this manner is so that the object filter can iterate the object and the trie at the same time, how would that work with a `bool contains()` function? I was explicitly trying to avoid having to generate and pass a vector to be validated by the trie.